File: akka/02-pingpong/java/src/main/java/PingPong.java
import akka.actor.AbstractActor;
import akka.actor.ActorRef;
import akka.actor.ActorSystem;
import akka.actor.Props;

public class PingPong {
	static ActorSystem actorSystem = ActorSystem.create("system");

	public static void main(String[] args) {
		final ActorRef ping = actorSystem.actorOf(Ping.props(), "ping");
		final ActorRef pong = actorSystem.actorOf(Pong.props(), "pong");
		ping.tell(new Ping.Message(), pong);
	}

	static void pause(long millis) {
		try {
			Thread.sleep(millis);
		} catch (InterruptedException ie) {
		}
	}
}

class Ping extends AbstractActor {
	public static class Message {
	}

	public static Props props() {
		return Props.create(Ping.class);
	}

	@Override
	public Receive createReceive() {
		return receiveBuilder().match(Message.class, this::ping).build();
	}

	private void ping(Message message) {
		System.out.println("Ping...");
		PingPong.pause(200);
		sender().tell(new Pong.Message(), self());
	}
}

class Pong extends AbstractActor {
	private static int counter;

	public static class Message {
	}

	public static Props props() {
		return Props.create(Pong.class);
	}

	@Override
	public Receive createReceive() {
		return receiveBuilder().match(Message.class, this::pong).build();
	}

	private void pong(Message message) {
		counter++;
		System.out.println("Pong! [" + counter + "]");
		if (counter >= 42) {
			PingPong.actorSystem.terminate();
			return;
		}
		PingPong.pause(200);
		sender().tell(new Ping.Message(), self());
	}
}

File: akka/02-pingpong/scala/PingPong.scala
import akka.actor.{ Actor, ActorLogging, ActorRef, ActorSystem, Props }

object Main extends App {
	val actorSystem = ActorSystem("system")
	var ping = actorSystem.actorOf(Props[Ping], name = "ping")
	var pong = actorSystem.actorOf(Props[Pong], name = "pong")
	ping.tell(Ping.Message, pong)
}

class Ping extends Actor {
	def receive = {
		case Ping.Message => 
			println("Ping...");
			Thread.sleep(200);
			sender.tell(Pong.Message, self)
	}
}

object Ping {
	case object Message
}

class Pong extends Actor {
	def receive = {
		case Pong.Message => 
			Pong.counter += 1
			println("Pong! [" + Pong.counter + "]");
			if (Pong.counter >= 42) {
				Main.actorSystem.terminate();
			} else {
				Thread.sleep(200);
				sender.tell(Ping.Message, self)
			}
	}
}

object Pong {
	var counter = 0
	case object Message
}

